lib/core/token.js

```diff
@@ -0,0 +1,246 @@
+        throw error
+      }
+    }
+
+    // Set token expiration
+    return this._setExpiration(tokenExpiration || false)
+  }
+
+  _setToken (token) {
+    const _key = this.$auth.options.token.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.setUniversal(_key, token)
+  }
+
+  _syncToken () {
+    const _key = this.$auth.options.token.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.syncUniversal(_key)
+  }
+
+  get () {
+    const _key = this.$auth.options.token.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.getUniversal(_key)
+  }
+
+  set (tokenValue) {
+    const token = addTokenPrefix(tokenValue, this.$auth.strategy.options.token.type)
+
+    this._setToken(token)
+    this._setHeader(token)
+    this._updateExpiration(token)
+
+    return token
+  }
+
+  sync () {
+    const token = this._syncToken()
+    this._syncExpiration()
+    this._setHeader(token)
+
+    return token
+  }
+
+  reset () {
+    this._clearHeader()
+    this._setToken(false)
+    this._setExpiration(false)
+  }
+
+  getStatus () {
+    return new TokenStatus(this.get(), this._getExpiration())
+  }
+
+  refreshAt () {
+    return (this._getExpiration() - Date.now()) * 0.75
+  }
+}
+
+export class RefreshToken {
+  constructor (auth) {
+    this.$auth = auth
+  }
+
+  _getExpiration () {
+    const _key = this.$auth.options.refreshTokenExpiration.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.getUniversal(_key)
+  }
+
+  _setExpiration (expiration) {
+    const _key = this.$auth.options.refreshTokenExpiration.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.setUniversal(_key, expiration)
+  }
+
+  _syncExpiration () {
+    const _key = this.$auth.options.refreshTokenExpiration.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.syncUniversal(_key)
+  }
+
+  _updateExpiration (refreshToken) {
+    let refreshTokenExpiration
+    const _tokenIssuedAtMillis = Date.now()
+    const _tokenTTLMillis = this.$auth.strategy.options.refreshToken.maxAge * 1000
+    const _tokenExpiresAtMillis = _tokenIssuedAtMillis + _tokenTTLMillis
+
+    try {
+      refreshTokenExpiration = jwtDecode(refreshToken).exp * 1000 || _tokenExpiresAtMillis
+    } catch (error) {
+      // If the token is not jwt, we can't decode and refresh it, use _tokenExpiresAt value
+      refreshTokenExpiration = _tokenExpiresAtMillis
+
+      if (!(error instanceof InvalidTokenError)) {
+        throw error
+      }
+    }
+
+    // Set token expiration
+    return this._setExpiration(refreshTokenExpiration || false)
+  }
+
+  _setToken (refreshToken) {
+    const _key = this.$auth.options.refreshToken.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.setUniversal(_key, refreshToken)
+  }
+
+  _syncToken () {
+    const _key = this.$auth.options.refreshToken.prefix + this.$auth.strategy.name
+
+    return this.$auth.$storage.syncUniversal(_key)
+  }
+
+  get () {
```

Comment:
get is little bit unclear for public interface (Suggestion: getToken or getRefreshToken)

Reply:
Even though i get your point, I strongly advise going for this interface:
```js
$auth.token.get()
$auth.token.set()
$auth.token.sync()
$auth.token.refresh()
// and
$auth.refreshToken.get()
$auth.refreshToken.set()
$auth.refreshToken.sync()
$auth.refreshToken.refresh()
// and
$auth.refreshTokens() // Plural
```
As this results in a very clean and clear interface.